File: frontend/src/domain/asset-state.ts
import { Schema } from "effect"

export const AssetStateSchema = Schema.Struct({
  state: Schema.Literal("active", "warning", "mint_blocked", "data_unavailable", "wind_down"),
  reason: Schema.String,
  reserveCoveragePercent: Schema.NullOr(Schema.Number),
  evidenceAtUnixMs: Schema.NullOr(Schema.Number),
  policyVersion: Schema.String,
  updatedAtUnixMs: Schema.Number,
})

export type AssetState = typeof AssetStateSchema.Type

export const assetStateLabel = (state: AssetState["state"]): string => ({
  active: "Aktywny",
  warning: "Ostrzeżenie",
  mint_blocked: "Emisja zablokowana",
  data_unavailable: "Brak danych",
  wind_down: "Wygaszanie",
})[state]

File: backend/src/domain/asset_state.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AssetStateCode {
    Active,
    Warning,
    MintBlocked,
    DataUnavailable,
    WindDown,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AssetState {
    pub state: AssetStateCode,
    pub reason: String,
    pub reserve_coverage_percent: Option<f64>,
    pub evidence_at_unix_ms: Option<u64>,
    pub policy_version: String,
    pub updated_at_unix_ms: u64,
}
